dig.go

```diff
@@ -246,28 +277,34 @@ func (cv connectionVisitor) Visit(res result) resultVisitor {
 
 	path := strings.Join(cv.currentResultPath, ".")
 
-	r, ok := res.(resultSingle)
-	if !ok {
-		return cv
-	}
+	switch r := res.(type) {
+	case resultSingle:
+		k := key{name: r.Name, t: r.Type}
 
-	k := key{name: r.Name, t: r.Type}
+		if conflict, ok := cv.keyPaths[k]; ok {
+			*cv.err = fmt.Errorf(
+				"cannot provide %v from %v in constructor %v: already provided by %v",
+				k, path, cv.n.ctype, conflict)
+			return nil
+		}
 
-	if conflict, ok := cv.keyPaths[k]; ok {
-		*cv.err = fmt.Errorf(
-			"cannot provide %v from %v in constructor %v: already provided by %v",
-			k, path, cv.n.ctype, conflict)
-		return nil
-	}
+		if _, ok := cv.c.providers[k]; ok {
+			*cv.err = fmt.Errorf(
+				"cannot provide %v from %v in constructor %v: already in the container",
```

Comment:
Hmmm...the `%v` of the conflict wasn't super-great for debugging, but it's a bummer to not have *any* identifying information about the other provider of the type.